File: templates/production-adapters/opencode-http-session/python/runner_replay.py
from __future__ import annotations

import json
import os
from pathlib import Path

from contracts import load_bundle_contract
from opencode_http import OpencodeHttpClient, OpencodeHttpConfig
from opencode_server import ensure_server, stop_server, write_server_state
from runner import DEFAULT_OPENCODE_URL, DEFAULT_WORKFLOW, REPO_ROOT, run_gate, run_stage
from runner_cli import parse_replay_args
from session_store import SessionStore


def main() -> int:
    args = parse_replay_args(workflow_default=DEFAULT_WORKFLOW, opencode_url_default=DEFAULT_OPENCODE_URL)
    contract = load_bundle_contract(args.workflow_path)
    if args.mode == "inspect":
        print(inspect_run_directory(args.run_dir))
        return 0
    if args.mode == "sessions":
        print(print_active_session_ids(args.run_dir))
        return 0
    if args.mode == "gate":
        return run_gate(run_dir=args.run_dir, gate=args.gate or "", contract=contract).exit_code
    store = SessionStore(args.run_dir)
    server = ensure_server(
        base_url=args.opencode_url,
        host=args.opencode_host,
        port=args.opencode_port,
        no_start=args.no_start_opencode,
        timeout_sec=args.agent_timeout_sec,
        run_dir=args.run_dir,
        workdir=REPO_ROOT,
        username=os.environ.get("OPENCODE_SERVER_USERNAME"),
        password=os.environ.get("OPENCODE_SERVER_PASSWORD"),
    )
    write_server_state(args.run_dir / "artifacts" / "opencode-server.json", server)
    client = OpencodeHttpClient(
        OpencodeHttpConfig(
            server.base_url,
            args.agent_timeout_sec,
            os.environ.get("OPENCODE_SERVER_USERNAME"),
            os.environ.get("OPENCODE_SERVER_PASSWORD"),
        )
    )
    try:
        run_stage(
            run_dir=args.run_dir,
            stage=args.stage or "",
            server_url=server.base_url,
            agent_timeout_sec=args.agent_timeout_sec,
            contract=contract,
            store=store,
            client=client,
        )
        return 0
    finally:
        stop_server(server)


def inspect_run_directory(run_dir: Path) -> str:
    artifacts_dir = run_dir / "artifacts"
    input_path = artifacts_dir / "input.json"
    session_path = artifacts_dir / "opencode-sessions.json"
    attach_path = artifacts_dir / "opencode-attach-commands.txt"
    runtime_state = {}
    sessions = {}
    if session_path.is_file():
        payload = json.loads(session_path.read_text(encoding="utf-8"))
        if isinstance(payload, dict):
            runtime_state = payload.get("runtimeState", {})
            sessions = payload.get("sessions", {})
    return "\n".join(
        [
            f"runDirectory={run_dir}",
            f"inputArtifact={input_path}",
            f"attachGuide={attach_path}",
            f"runtimeState={json.dumps(runtime_state, ensure_ascii=False, sort_keys=True)}",
            f"sessions={json.dumps(sessions, ensure_ascii=False, sort_keys=True)}",
        ]
    )


def print_active_session_ids(run_dir: Path) -> str:
    session_path = run_dir / "artifacts" / "opencode-sessions.json"
    if not session_path.is_file():
        return ""
    payload = json.loads(session_path.read_text(encoding="utf-8"))
    sessions = payload.get("sessions", {}) if isinstance(payload, dict) else {}
    if not isinstance(sessions, dict):
        return ""
    lines: list[str] = []
    for stage in sorted(sessions):
        item = sessions.get(stage)
        if isinstance(item, dict) and isinstance(item.get("id"), str):
            lines.append(f"{stage}={item['id']}")
    return "\n".join(lines)


if __name__ == "__main__":
    raise SystemExit(main())

File: templates/production-adapters/opencode-http-session/typescript/runner-replay.ts
import { join } from "node:path"

import { loadBundleContract } from "./contracts"
import { readJsonObject } from "./json-store"
import { OpencodeHttpClient } from "./opencode-http"
import { ensureServer, stopServer, writeServerState } from "./opencode-server"
import { DEFAULT_OPENCODE_URL, DEFAULT_WORKFLOW, REPO_ROOT, runGate, runStage } from "./runner"
import { parseReplayArgs } from "./runner-cli"
import { SessionStore } from "./session-store"

async function main(): Promise<number> {
  const args = parseReplayArgs(Bun.argv.slice(2), { opencodeUrl: DEFAULT_OPENCODE_URL, workflowPath: DEFAULT_WORKFLOW })
  const bundle = await loadBundleContract(args.workflowPath)
  if (args.mode === "inspect") {
    console.log(await inspectRunDirectory(args.runDirectory))
    return 0
  }
  if (args.mode === "sessions") {
    console.log(await printActiveSessionIds(args.runDirectory))
    return 0
  }
  if (args.mode === "gate") {
    return (await runGate({ bundle, gate: args.gate ?? "", runDirectory: args.runDirectory })).exitCode
  }
  const store = new SessionStore(args.runDirectory)
  const server = await ensureServer({
    baseUrl: args.opencodeUrl,
    host: args.opencodeHost,
    noStart: args.noStartOpencode,
    password: process.env.OPENCODE_SERVER_PASSWORD,
    port: args.opencodePort,
    runDirectory: args.runDirectory,
    timeoutMs: args.agentTimeoutMs,
    username: process.env.OPENCODE_SERVER_USERNAME,
    workdir: REPO_ROOT,
  })
  await writeServerState(join(args.runDirectory, "artifacts", "opencode-server.json"), server)
  const client = new OpencodeHttpClient({
    baseUrl: server.baseUrl,
    password: process.env.OPENCODE_SERVER_PASSWORD,
    timeoutMs: args.agentTimeoutMs,
    username: process.env.OPENCODE_SERVER_USERNAME,
  })
  try {
    await runStage({
      agentTimeoutMs: args.agentTimeoutMs,
      bundle,
      client,
      runDirectory: args.runDirectory,
      serverUrl: server.baseUrl,
      stage: args.stage ?? "",
      store,
    })
    return 0
  } finally {
    stopServer(server)
  }
}

async function inspectRunDirectory(runDirectory: string): Promise<string> {
  const sessionPath = join(runDirectory, "artifacts", "opencode-sessions.json")
  const payload = (await Bun.file(sessionPath).exists()) ? await readJsonObject(sessionPath) : {}
  return [
    `runDirectory=${runDirectory}`,
    `inputArtifact=${join(runDirectory, "artifacts", "input.json")}`,
    `attachGuide=${join(runDirectory, "artifacts", "opencode-attach-commands.txt")}`,
    `runtimeState=${JSON.stringify(payload.runtimeState ?? {})}`,
    `sessions=${JSON.stringify(payload.sessions ?? {})}`,
  ].join("\n")
}

async function printActiveSessionIds(runDirectory: string): Promise<string> {
  const sessionPath = join(runDirectory, "artifacts", "opencode-sessions.json")
  if (!(await Bun.file(sessionPath).exists())) {
    return ""
  }
  const payload = await readJsonObject(sessionPath)
  const sessions = payload.sessions
  if (typeof sessions !== "object" || sessions === null || Array.isArray(sessions)) {
    return ""
  }
  return Object.entries(sessions as Record<string, Record<string, unknown>>)
    .filter(([, item]) => typeof item.id === "string")
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([stage, item]) => `${stage}=${item.id}`)
    .join("\n")
}

if (import.meta.main) {
  process.exit(await main())
}
